src/Avalonia.Native/IAvnMenu.cs

```diff
@@ -0,0 +1,176 @@
+        }
+
+        internal NativeMenu ManagedMenu { get; private set; }
+
+        public static IAvnMenu Create(IAvaloniaNativeFactory factory)
+        {
+            var events = new MenuEvents();
+
+            var menu = factory.CreateMenu(events);
+
+            events.Initialise(menu);
+
+            menu._events = events;
+
+            return menu;
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                _events.Dispose();
+            }
+        }
+
+        private void RemoveAndDispose(IAvnMenuItem item)
+        {
+            _menuItemLookup.Remove(item.ManagedMenuItem);
+            _menuItems.Remove(item);
+            RemoveItem(item);
+
+            item.Deinitialise();
+            item.Dispose();
+        }
+
+        private void MoveExistingTo(int index, IAvnMenuItem item)
+        {
+            _menuItems.Remove(item);
+            _menuItems.Insert(index, item);
+
+            RemoveItem(item);
+            InsertItem(index, item);
+        }
+
+        private IAvnMenuItem CreateNewAt(IAvaloniaNativeFactory factory, int index, NativeMenuItemBase item)
+        {
+            var result = CreateNew(factory, item);
+
+            result.Initialise(item);
+
+            _menuItemLookup.Add(result.ManagedMenuItem, result);
+            _menuItems.Insert(index, result);
+
+            InsertItem(index, result);
+
+            return result;
+        }
+
+        private IAvnMenuItem CreateNew(IAvaloniaNativeFactory factory, NativeMenuItemBase item)
+        {
+            var nativeItem = item is NativeMenuItemSeperator ? factory.CreateMenuItemSeperator() : factory.CreateMenuItem();
+            nativeItem.ManagedMenuItem = item;
+
+            return nativeItem;
+        }
+
+        internal void Initialise(AvaloniaNativeMenuExporter exporter, NativeMenu managedMenu, string title)
+        {
+            _exporter = exporter;
+            ManagedMenu = managedMenu;
+
+            ((INotifyCollectionChanged)ManagedMenu.Items).CollectionChanged += OnMenuItemsChanged;
+
+            if (!string.IsNullOrWhiteSpace(title))
+            {
+                using (var buffer = new Utf8Buffer(title))
+                {
+                    Title = buffer.DangerousGetHandle();
+                }
+            }
+        }
+
+        internal void Deinitialise()
+        {
+            ((INotifyCollectionChanged)ManagedMenu.Items).CollectionChanged -= OnMenuItemsChanged;
+
+            foreach (var item in _menuItems)
+            {
+                item.Deinitialise();
+                item.Dispose();
+            }
+        }
+
+        internal void Update(IAvaloniaNativeFactory factory, NativeMenu menu)
+        {
+            if (menu != ManagedMenu)
+            {
+                throw new ArgumentException("The menu being updated does not match.", nameof(menu));
+            }
+
+            for (int i = 0; i < menu.Items.Count; i++)
```

Comment:
How does it works with toplevel menu? The item at index 0 should be the application menu, which is managed by the native part. So for toplevel menus we should use an offset for menu items managed by the exporter.

Am I missing something here?

Reply:
https://github.com/AvaloniaUI/Avalonia/blob/4341caef1a615a3dc407a2dabc6981aa7c46b8aa/native/Avalonia.Native/src/OSX/window.mm#L1325

We should probably add a native call for checking if the item at index 0 is the app menu and act accordingly.

Reply:
I beleive this is resolved now with index translation based on flag?